### ISSUE TITLE
Create a widget compatible with other Jupyter notebook widgets

### ISSUE THREAD

**Issue by ejeschke** [ejeschke/ginga#497]
Ginga already has a web browser back end (see #168) and there is a Ginga [viewer that is tailored for working in/with Jupyter notebooks](https://github.com/ejeschke/ginga/blob/master/ginga/examples/jupyter-notebook/ginga_ipython_demo.ipynb).

This issue is exploring whether a native Jupyter widget can be used as the back end of a Ginga viewer.  I've created a new `ginga.web.jupyterw` back end module.  It uses the Jupyter `Image` widget as the surface and subclasses the PIL backend for the rendering.  You can see some initial experiments of the use in [this notebook](https://github.com/ejeschke/ginga/blob/master/ginga/examples/jupyter-notebook/Jupyter%20Widget%20Ideas.ipynb).

The most useful feature of this version is that (as the example shows) the widget can be packed in with other Jupyter widgets and the packing and interaction between the widgets all just works.

The downside is that currently there is no support for UI events in a Jupyter image widget--therefore there are no cursor or keyboard interactions possible with the widget. Everything must be controlled by other widgets or programmatically.

**Comment by ejeschke** [comment]
ping @mwcraig, @austinbeauch

**Comment by austinbeauch** [comment]
Hi @ejeschke, took a bit of a break but now I'm back to working on the widgets. 

Are any of the widgets compatible with an EnhancedCanvasView type? I've tried implementing the strategies from [examples/pg/example2_pg.py](https://github.com/ejeschke/ginga/tree/master/ginga/examples/pg) which is referenced in #168 but that file is doesn't use a notebook and inserts the widgets using a TopLevel object. I've tried using but it just won't seem to go.

Do you know of a way to insert one of ginga.web.pgw Widget's on a canvas that's been created like [here?](https://github.com/ejeschke/ginga/blob/master/ginga/examples/jupyter-notebook/ginga_ipython_demo.ipynb)

Perhaps another approach would be to embed the viewing window in the notebook and then use ipywidgets.

**Comment by ejeschke** [comment]
> Perhaps another approach would be to embed the viewing window in the notebook and then use ipywidgets.

@austinbeauch, that might be not too difficult.  Check out the output of the `embed()` call, it may be something that you can wrap using one of the Jupyter widgets (like a nested HTML item or something).

I think trying to embed the ginga wrapped widgets (from the `example2_pg.py`, etc) in the notebook will ultimately be unsatisfactory. You are then kind of mixing apples and oranges.  It seems to me that you really want something that implements a Jupyter compatible widget, because then (as described above) you can just mix it compatibly with other widgets in the Jupyter ecosystem.

Did you take a look at the code (`ginga.web.jupyterw`) and example notebook listed in the top of this issue?  If you can figure out a way to connect pointer and keyboard event callbacks via Jupyter's `Image` widget it would be the best way to implement what you are looking for.

**Comment by austinbeauch** [comment]
We got it working! Just created a new subclass of `ipg.EnhancedCanvasView` and then set it as the `viewer_class` when creating the server from `ipg.make_server`. Probably not how it was designed to be used but it seems to be performing correctly (for now).

**Comment by ejeschke** [comment]
Hey, that's great!  I think it sounds like you are using the `viewer_class` mechanism as it was intended (it is designed so that you can create new kinds of viewers by subclassing).

Can you attach a screenshot of what you have working?

**Comment by austinbeauch** [comment]
Here's a screenshot of the class. We're still working with the accept/reject buttons so they don't have much functionality right now. The method near at the bottom that deals with "candidates" is part of our program so no need to worry about that. And the line which sets that as the viewer class is 
`self.web_server = ipg.make_server(host=self.host, 
  port=self.port, use_opencv=False, viewer_class=self.viewer_class)`

![screenshot from 2017-06-20 09-47-54](https://user-images.githubusercontent.com/18565932/27345173-8e64dcd6-559d-11e7-82df-fa092e2f1b24.png)

**Comment by ejeschke** [comment]
I meant a screenshot of your GUI!  :grin:

**Comment by ejeschke** [comment]
But, code looks good!

**Comment by austinbeauch** [comment]
Whoops... now you have both! Grey box is the viewer/canvas, didn't load any image into it. Are there any GUI competitions? I think this could be one of the best ever created.

![screenshot from 2017-06-20 12-17-10](https://user-images.githubusercontent.com/18565932/27351489-666f2cb2-55b2-11e7-807b-0eecd7aab93f.png)

**Comment by pllim** [comment]
Re: Competition -- Not really, but there is always Google Summer of Code for next year (2018)! 😉

**Comment by mwcraig** [comment]
@austinbeauch --this looks promising!  Are you contructing the widgets using Jupyter's ipywidgets, or ginga's  widget set?

**Comment by austinbeauch** [comment]
@mwcraig Completely from ginga's widget set, I haven't touched ipywidgets at all! That viewer/gui opens up in a new tab, so it's separate from the ipython command shell.

**Comment by ejeschke** [comment]
> Grey box is the viewer/canvas, didn't load any image into it. Are there any GUI competitions? I think this could be one of the best ever created.

You have to load an impressive astro image and overlay a canvas with some graphics!

**Comment by pllim** [comment]
> impressive astro image

An image of space cats is always an option...

**Comment by mwcraig** [comment]
https://github.com/eteq/catterplotpy

**Comment by mwcraig** [comment]
@ejeschke -- I think a better approach, rather trying to add events to the Image widget (though I'll ask about that at scipy) is, in the short run, to put the viewer in an IFrame in a jupyter HTML widget. See https://github.com/glowing-waffle/glowing-waffles/blob/master/glowing_waffles/visualization/astro_widgets.py (and please ignore the package name, which won't stick around much longer).

I just pushed a few changes to take advantage of your recent modifications, and it seems to work with some glitches (e.g. I'm not handling ports right, the widget really doesn't resize nicely).

In the longer run I think a widget that runs your javascript but communicates through the widget framework rather than a tornado server might be the way to go, though I have no idea how convoluted that is.

Sample notebook code to use it is:

```python 
from astro_widget import ImageViewer
i = ImageViewer(show_position=True)
i
```

This is more at the proof-of-concept stage still...

**Comment by mwcraig** [comment]
One question -- what is the right way to make multiple viewers if one wants multiple viewers?

**Comment by mwcraig** [comment]
Never mind that last question, the issue is that I shouldn't be creating multiple servers....

**Comment by ejeschke** [comment]
> what is the right way to make multiple viewers if one wants multiple viewers?

@mwcraig, in the original code you could just call `get_viewer()` on the server with another name and you'd get another viewer.  As you say, just create one server and you can create many viewer windows from that.

**Comment by ejeschke** [comment]
Looking at the code, maybe you could make a factory class that generates viewers, that would encapsulate the server creation and pass in a separate string for each new viewer that was created.  Like so:
```python
vf = ViewerFactory(...)

v1 = vf.get_viewer()
v2 = vf.get_viewer()
...
```

**Comment by mwcraig** [comment]
Thanks, @ejeschke, I'll give that a try!

**Comment by mwcraig** [comment]
@ejeschke -- Talked to @jasongrout at scipy about jupyterlab/widgets, a fits viewer and a jupyter widget. Pinging here then getting out of the way.

**Comment by jasongrout** [comment]
@ejeschke - happy to help - it would be pretty easy to add some notion of click or selection to the image widget, or even better, subclass the image widget and add that interaction in a separate widget.

**Comment by ejeschke** [comment]
@jasongrout, I like the idea of using a subclass of the image widget.  Are there any code examples of how to set up event callbacks for the widget subclass?  We would be looking for:
- keypresses (and releases, if possible)
- "pointer" clicks (left and right)
- scrolling

**Comment by mwcraig** [comment]
The code mentioned in [this gitter comment on bqplot](https://gitter.im/bloomberg/bqplot?at=596a5f0ef5b3458e304d4949) might be useful as a starting point...

**Comment by ejeschke** [comment]
@mwcraig, I'm on vacation for another week & will take a look at that when I get back, but do you want to take a stab at adding some callbacks to the image widget?

**Comment by jasongrout** [comment]
@ejeschke - sorry, I missed your comment above. A widget has both a javascript and a kernel object instance. These objects can sync state and also send custom messages to each other. The button control has an example of using custom messages to trigger callbacks on click.

* On the javascript side, the view's `.send` method is called to send a custom message to the kernel side: https://github.com/jupyter-widgets/ipywidgets/blob/f6656744c9407df8d8101ef0bfeaf62dbda6a736/packages/controls/src/widget_button.ts#L117
* On the kernel side, the button class registers a handler for custom messages, and uses that to trigger a callback when a clicked message is received: https://github.com/jupyter-widgets/ipywidgets/blob/f6656744c9407df8d8101ef0bfeaf62dbda6a736/ipywidgets/widgets/widget_button.py#L63

**Comment by mwcraig** [comment]
@ejeschke -- I'll try to take a look, but I'm on the road for the next couple of weeks (and giving a talk this Monday I should really start writing 😱), so I'm not optimistic...

**Comment by ejeschke** [comment]
Merged in support in #562. Closing.
